numba/ir_utils.py

```diff
@@ -1854,3 +1854,68 @@ def raise_on_unsupported_feature(func_ir):
                "nopython-mode\n\nConflicting calls found at:\n %s")
         buf = '\n'.join([x.strformat() for x in gdb_calls])
         raise UnsupportedError(msg % buf)
+
+
+class InlineInlinables(object):
+
+    def __init__(self, func_ir, recursive=False):
+        self.func_ir = func_ir
+        self.recursive = recursive
+
+    def run(self):
+        """Run inlining of inlinables
+        """
+        if config.DEBUG:
+            print('before inline'.center(80, '-'))
+            print(self.func_ir.dump())
+            print(''.center(80, '-'))
+        modified = False
+        work_list = list(self.func_ir.blocks.items())
+        while work_list:
+            label, block = work_list.pop()
+            for i, instr in enumerate(block.body):
+                if isinstance(instr, ir.Assign):
+                    expr = instr.value
+                    if isinstance(expr, ir.Expr) and expr.op == 'call':
+                        if guard(self._do_work, work_list, block, i, expr):
+                            modified = True
+                            break # because block structure changed
+
+        if modified:
+            remove_dels(self.func_ir.blocks)
+            # repeat dead code elimination until nothing can be further
+            # removed
+            while (remove_dead(self.func_ir.blocks, self.func_ir.arg_names,
+                               self.func_ir)):
+                pass
+        if config.DEBUG:
+            print('after inline'.center(80, '-'))
+            print(self.func_ir.dump())
+            print(''.center(80, '-'))
+
+    def _do_work(self, work_list, block, i, expr):
+        from numba.inline_closurecall import inline_closure_call
+        to_inline = self.func_ir.get_definition(expr.func)
+        # do not handle closure inlining here, another pass deals with that.
+        if getattr(to_inline, 'op', False) == 'make_function':
+            return False
+        val = getattr(to_inline, 'value', False)
```

Comment:
I had to checkout the PR and put a breakpoint here to understand this line. I suggest `isinstance(to_inline, ir.Global)` to make it clear.
I suggest using `require` in guarded functions to avoid these branches. For example, `require(isinstance(to_inline, ir.Global))` will avoid the branch.